frontend/lib/src/hocs/withExpandable/withExpandable.tsx

```diff
@@ -49,25 +52,50 @@ function withExpandable<P>(
   const ExpandableComponent = (props: WrapperProps<P>): ReactElement => {
     const {
       label,
-      expanded: initialExpanded,
+      expandableState,
       empty,
       widgetsDisabled,
       isStale,
       ...componentProps
     } = props
 
-    const [expanded, setExpanded] = useState<boolean>(initialExpanded)
+    const [expanded, setExpanded] = useState<boolean>(
+      expandableState != ExpandableState.COLLAPSED
+    )
+
+    // useEffect(() => {
+    //   setExpanded(initialExpanded)
+    //   // Having `label` in the dependency array here is necessary because
+    //   // sometimes two distinct expanders look so similar that even the react
+    //   // diffing algorithm decides that they're the same element with updated
+    //   // props (this happens when something in the app removes one expander and
+    //   // replaces it with another in the same position).
+    //   //
+    //   // By adding `label` as a dependency, we ensure that we reset the
+    //   // expander's `expanded` state in this edge case.
+    // }, [label, initialExpanded])
+
     useEffect(() => {
-      setExpanded(initialExpanded)
-      // Having `label` in the dependency array here is necessary because
-      // sometimes two distinct expanders look so similar that even the react
-      // diffing algorithm decides that they're the same element with updated
-      // props (this happens when something in the app removes one expander and
-      // replaces it with another in the same position).
-      //
-      // By adding `label` as a dependency, we ensure that we reset the
-      // expander's `expanded` state in this edge case.
-    }, [label, initialExpanded])
+      let timer: NodeJS.Timeout | number | undefined
+
+      switch (expandableState) {
+        case ExpandableState.EXPANDED:
+          setExpanded(true)
+          break
+        case ExpandableState.COLLAPSED:
+          setExpanded(false)
+          break
+        case ExpandableState.AUTO_COLLAPSED:
+          timer = setTimeout(() => setExpanded(false), AUTO_COLLAPSE_DELAY)
```

Comment:
There's a slightly weird behavior with this: if the expander is collapsed when this timer starts and then expanded (such as by the user) before it completes, the timer will fire and collapse it again.

I think it's better if this behavior doesn't happen. I think it could be fixed here if we check the current expanded state and only start the timer if it's currently expanded? Thoughts?

https://github.com/streamlit/streamlit/assets/116604821/64852428-a802-47c2-867a-f8de03add5fd

Reply:
Good point, I will change this to only use the timer if it is expanded.